File: leetcode/python/49_group_anagrams.py
from string import ascii_lowercase
from typing import List, Dict, Tuple


class Solution:

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        mapping = {}  # type: Dict[Tuple[int, ...], int]
        groups = []  # type: List[List[str]]
        for s in strs:
            letter_count = self.get_letter_count(s)
            group_index = mapping.setdefault(letter_count, len(groups))
            if group_index >= len(groups):
                groups.append([s])
            else:
                groups[group_index].append(s)
        return groups

    def get_letter_count(self, s: str) -> Tuple[int, ...]:
        letter_count = [0] * len(ascii_lowercase)
        for ch in s:
            letter_count[ord(ch) - ord(ascii_lowercase[0])] += 1
        return tuple(letter_count)

File: leetcode/java/49GroupAnagrams.java
/**
 * T: O(N) S: O(N)
 *
 * <p>Given that lengths of the elements are low, we can sort them to get a common representation
 * for an anagram. After that, all of anagrams will map to the same representation - collect them
 * and return at the end.
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
    Map<String, List<String>> groups = new HashMap<>();
    for (String str : strs) {
      char[] chars = str.toCharArray();
      Arrays.sort(chars);
      String key = new String(chars);
      if (!groups.containsKey(key)) {
        groups.put(key, new ArrayList<String>());
      }
      List<String> existingValues = groups.get(key);
      existingValues.add(str);
      groups.put(key, existingValues);
    }
    return new ArrayList<List<String>>(groups.values());
  }
}
